servers/src/main/java/tachyon/worker/file/FileDataManager.java

```diff
@@ -75,13 +77,20 @@ public synchronized void persistFile(long fileId, List<Long> blockIds) throws IO
     OutputStream outputStream = mUfs.create(dstPath);
     final WritableByteChannel outputChannel = Channels.newChannel(outputStream);
 
+    Map<Long, Long> blockIdToLockId = Maps.newHashMap();
+    // lock all the blocks to prevent any eviction
     for (long blockId : blockIds) {
       long lockId;
       try {
         lockId = mBlockDataManager.lockBlock(Sessions.CHECKPOINT_SESSION_ID, blockId);
+        blockIdToLockId.put(blockId, lockId);
       } catch (BlockDoesNotExistException e) {
         throw new IOException(e);
       }
+    }
```

Comment:
a general way to solve this is to always acquire locks in the same order, and release all previously grabbed locks if the attempt to acquire the next lock failed. http://herbsutter.com/2007/12/11/effective-concurrency-use-lock-hierarchies-to-avoid-deadlock/